flashlight/fl/dataset/SpanDataset.cpp

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <stdexcept>
+
+#include "flashlight/fl/dataset/SpanDataset.h"
+
+namespace fl {
+SpanDataset::SpanDataset(
+    std::shared_ptr<const Dataset> dataset,
+    const int64_t offset,
+    const int64_t length)
+    : dataset_(dataset), offset_(offset) {
+  size_ = (length < 0) ? (dataset_->size() - offset_) : length;
+  if (size_ + offset_ > dataset_->size()) {
+    throw std::out_of_range(
+        "Dataset length out of range (larger than underlying dataset)");
+  }
+}
+
+std::vector<Tensor> SpanDataset::get(const int64_t idx) const {
+  checkIndexBounds(idx);
+
+  std::vector<Tensor> result;
```

Comment:
small optimization: we can reserve or preemptively size `result` with `f.size()`

Reply:
oh can I just return `dataset_->get(idx + offset_);`?